Adding documentation to create dask DataFrame from HDF5

linwoodc3 opened the pull request:
I wrote this recommendation up in issue #664, but wanted to put the same write up here.  The bottomline is, I'd like to add a small paragraph of documentation to the [`dask` DataFrame creation](http://dask.pydata.org/en/latest/dataframe.html) section.  

For [excruciating detail (with screen shots), please read the linked issue](https://github.com/dask/dask/pull/664#issuecomment-234730449), but in short, this pull request should help marry the [dask.dataframe.io.py](https://github.com/dask/dask/blob/master/dask/dataframe/io.py#L537-L664) comments up with the `dask` documentation.

linwoodc3 commented:
If anyone can help on why the "builds" are failing; i only changed documentation. Any help is appreciated..new to this.

nirizr commented:
This fails because code inside documentation is automatically run as a test and is expected to pass.
You can test documentation using `py.test --doctest-modules` (for further instructions on how to run tests, please consult contributing documentation).
You should either fix it or add a `# doctest: +SKIP` comment to the code line.

You might want to take a look at numpydoc or doctests.

P.S.
You don't need to add comments regarding the specific code examples like you did, it's better to add those as actual documentation text.

linwoodc3 commented:
Thanks @nirizr .

nirizr commented on docs/source/dataframe-create.rst:
```diff
@@ -58,29 +58,39 @@ and not actively maintained; use at your own risk.
-   >>> df = c.to_dask()
+   >>> from castra import Castra # doctest: +SKIP
+   >>> c = Castra(path='/my/castra/file') # doctest: +SKIP
+   >>> df = c.to_dask() # doctest: +SKIP
```
I don't think you need to add `# doctest: +SKIP` to every line in a code block. I think one is enough.

nirizr commented on docs/source/dataframe-create.rst:
```diff
@@ -58,29 +58,39 @@ and not actively maintained; use at your own risk.
+   >>> dset = f.create_dataset("mydataset", (1000000,), dtype='i')
+   >>> dset[...] = np.arange(1000000)
+   >>> print dset.shape
+   (1000000,)
```
The purpose of docstings is to be clear and simple, I think your original version with a +SKIP might be better.

nirizr commented on docs/source/dataframe-create.rst:
```diff
@@ -58,29 +58,39 @@ and not actively maintained; use at your own risk.
+   >>> import dask.dataframe as dd
+   >>> dd.read_hdf('myfile.1.hdf5', '/mydataset', chunksize=1000) # doctest: +SKIP
    
-   >>> # OR Load multiple hdf5 datasets into a dask DataFrame
```
My intention was not that you should remove those comments.
Instead, having the following code:
```
OR Load multiple hdf5 datasets into a dask DataFrame

>>> dd.read_hdf('myfile.1.hdf5', '/*', chunksize=1000000)
```

linwoodc3 replied:
I reverted back.  The original documentation kept failing the doctest; Here is a screen shot of the previous documentation failing; I had to add the `SKIP` to each line for it to pass.

![screen shot 2016-07-23 at 8 33 57 pm](https://cloud.githubusercontent.com/assets/2884914/17081039/ea5a34c0-5114-11e6-9068-b6778609be1c.png)

mrocklin commented:
Sorry for being absent here.  We actually don't test the HTML documentation at all, so the `# doctest: +SKIP` text is unnecessary.  We *do* test function docstrings though, so you would need to do this if you were, for example, changing the docstring of the `dd.read_hdf` function itself.  

The original failures look to be completely unrelated and were from anaconda.org being down when travis-ci was trying to download conda packages.

mrocklin commented on docs/source/dataframe-create.rst:
```diff
@@ -58,12 +58,29 @@ and not actively maintained; use at your own risk.
+From HDF5
+----------
+
+`HDF5 <https://www.hdfgroup.org/HDF5/doc/H5.intro.html>`_ is a Hierarchical Data Format (HDF) designed to store and organize large amounts of data.  Similar to the `pandas I\/O API <http://pandas.pydata.org/pandas-docs/stable/io.html>`_,    `dask <(http://dask.pydata.org/en/latest/index.html>`_ can create a DataFrame directly from `HDF5 <https://www.hdfgroup.org/HDF5/doc/H5.intro.html>`_ datasets ...
```
We should probably refer to the `dask.dataframe` module instead of `dask` here.  Otherwise people get the impression that Dask is just a dataframe library.

linwoodc3 replied:
Okay; will change it.

mrocklin commented on docs/source/dataframe-create.rst:
```diff
@@ -58,12 +58,29 @@ and not actively maintained; use at your own risk.
+   >>> # Load hdf5 into dask DataFrame 
+   >>> dd.read_hdf('myfile.1.hdf5', '/x', chunksize=1000000) # doctest: +SKIP
+   
+   
```
Perhaps shrink this down to single-line breaks between code snippets rather than double line breaks?

linwoodc3 replied:
Will do

mrocklin commented on docs/source/dataframe-create.rst:
```diff
@@ -58,12 +58,29 @@ and not actively maintained; use at your own risk.
+   
+   
+   >>> # OR Load multiple hdf5 datasets into a dask DataFrame 
+   >>> dd.read_hdf('myfile.1.hdf5', '/*', chunksize=1000000) # doctest: +SKIP
```
I agree with you that this section is important.  I recommend that we elevate its prominence in the docs by moving it just under the CSV section and above the Array section.

linwoodc3 replied:
Will do; I'll make that change and do a push and report back here once I make changes as you suggest.

linwoodc3 commented:
@mrocklin , I made the clean up and formatting as suggested.  Let me know if I need to change anything.  Here's a summary:

*  Moved HDF5 section just below the CSV section
* Removed double space breaks in code; broke into smaller sections in HDF5 section and removed comments from code
*  Added two introductory sentences to the Dask Create DataFrame page with cross reference to Dask DataFrame Overview page (fractal documentation...see below)
*  Added a cross reference to Dask Overview page (fractal documentation..see below)
* Added cross reference to Dask Array and Overview pages (fractal docmentation..see below)

This "fractal documentaion" concept comes from Jacob Kaplan-Moss of the Python Django project and is explained at point 21:05  thru 22:13 of this PyCon 2011 YouTube video (https://www.youtube.com/watch?v=z3fRu9pkuXE).

Thank you for the opportunity to contribute!

Respectfully

mrocklin commented:
This looks great to me and is, to my eye, very well written.  Thank you for the contribution.  I encourage you to keep it up if you have the time and will; our documentation was written hastily with a "lets just get this stuff down" motivation and could use a tasteful eye.

Merging.
